### PR TITLE
Add error checking if provided shape to tfrecord can house underlying data

### PR DESCRIPTION
**Pull request by JanuszL**
- when the user provides a shape to tfrecord feature descrption there was not check if the reader output tensor is big enough and there won't be out of bound access. This change adds proper DALI_ENFORCE

Signed-off-by: Janusz Lisiecki <jlisiecki@nvidia.com>

#### Why we need this PR?
*Pick one, remove the rest*
- It adds error checking if provided shape to tfrecord can house underlying data

#### What happened in this PR?
*Fill relevant points, put NA otherwise. Replace anything inside []*
 - What solution was applied:
     when the user provides a shape to tfrecord feature descrption there was not check if the reader output tensor is big enough and there won't be out of bound access. This change adds proper DALI_ENFORCE
 - Affected modules and functionalities:
     tfrecord_parser
 - Key points relevant for the review:
     NA
 - Validation and testing:
     new test is added
 - Documentation (including examples):
     NA

Relates to https://github.com/NVIDIA/DALI/issues/2701
**JIRA TASK**: *[NA]*

### CODE REVIEW

**Comment by JanuszL** [comment]
!build

File: dali/operators/reader/parser/tfrecord_parser.h [hunk 1]
```diff
@@ -74,11 +74,16 @@ class TFRecordParser : public Parser<Tensor<CPUBackend>> {
...
       }
+      ssize_t number_of_elms = 0;
       switch (f.GetType()) {
         case FeatureType::int64:
+          number_of_elms = encoded_feature.int64_list().value().size();
           if (!f.HasShape()) {
-            output.Resize(InferShape(f, encoded_feature.int64_list().value().size()));
+            output.Resize(InferShape(f, number_of_elms));
           }
+          DALI_ENFORCE(number_of_elms <= output.size(), make_string("Output tensor is too "
```
Comment (jantonguirao) [inline dali/operators/reader/parser/tfrecord_parser.h:84]:
```suggestion
                   DALI_ENFORCE(number_of_elms <= output.size(), make_string("Output tensor shape is too "
                        "small: [", output.shape(),
                        "]. Expected at least ", number_of_elms, " elements."));
```

File: dali/test/python/test_operator_index_reader.py [hunk 2]
```diff
@@ -83,4 +84,20 @@ def define_graph(self):
...
             assert np.array_equal(a.as_array(), b.as_array())
-        _ = pipe_org.run()
\ No newline at end of file
+        _ = pipe_org.run()
+
+def test_wrong_feature_shape():
+    features = {
+        'image/encoded': tfrec.FixedLenFeature((), tfrec.string, ""),
+        'image/object/bbox': tfrec.FixedLenFeature([], tfrec.float32, -1.0),
+        'image/object/class/label': tfrec.FixedLenFeature([], tfrec.int64, -1),
```
Comment (jantonguirao) [inline dali/test/python/test_operator_index_reader.py:93]: Can you add a comment explaining why an error is expected?

File: dali/operators/reader/parser/tfrecord_parser.h (hunk 1 above)
Comment (JanuszL) [reply dali/operators/reader/parser/tfrecord_parser.h:84]: Done

File: dali/test/python/test_operator_index_reader.py (hunk 2 above)
Comment (JanuszL) [reply dali/test/python/test_operator_index_reader.py:93]: Done

File: dali/test/python/test_operator_index_reader.py [hunk 3]
```diff
@@ -83,4 +84,20 @@ def define_graph(self):
...
+    }
+    test_dummy_data_path = os.path.join(get_dali_extra_path(), 'db', 'coco_dummy')
+    pipe = Pipeline(1, 1, 0)
+    with pipe:
+        input = fn.tfrecord_reader(path = os.path.join(test_dummy_data_path, 'small_coco.tfrecord'),
+                                   index_path = os.path.join(test_dummy_data_path, 'small_coco_index.idx'),
+                                   features = features)
+    pipe.set_outputs(input['image/encoded'], input['image/object/class/label'], input['image/object/bbox'])
+    pipe.build()
+    assert_raises(RuntimeError, pipe.run)
```
[the remark this answers is not in the document]
Comment (JanuszL) [reply dali/test/python/test_operator_index_reader.py]: Done

**Review by jantonguirao — APPROVED** [review]

**Review by awolant — APPROVED** [review]

**Comment by JanuszL** [comment]
!build

**Comment by JanuszL** [comment]
!build

**Pull request merged**